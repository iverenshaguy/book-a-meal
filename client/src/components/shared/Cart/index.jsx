import React, { Fragment, Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import ReactTooltip from 'react-tooltip';
import LinkBtn from '../Link';
import { orderItemPropTypes } from '../../../helpers/proptypes';
import checkShopOpen from '../../../helpers/checkShopOpen';
import calculateCashEarnedFromOrder from '../../../helpers/calculateCashEarnedFromOrder';
import CloseIcon from '../CloseIcon';
import './Cart.scss';

/**
 * @exports
 * @extends Component
 * @class Cart
 * @returns {JSX} Cart
 */
class Cart extends Component {
  static propTypes = {
    order: PropTypes.arrayOf(orderItemPropTypes).isRequired,
    handleQuantityInputChange: PropTypes.func.isRequired,
    removeOrderItem: PropTypes.func.isRequired
  }

  state = {
    show: false
  }

  /**
   * @memberof Cart
   * @returns {JSX} Cart Component
   */
  toggleCart = () => {
    this.setState(prevState => ({
      show: !prevState.show
    }));
  }

  /**
   * @memberof Cart
   * @returns {JSX} Checkout Btn
   */
  renderCheckoutBtn = () => {
    const { order } = this.props;
    const disabled = order.length === 0 ? 'disabled' : null;

    return (
      <Link to="/order-review">
        <button type="button" className="btn btn-pri btn-block checkout-btn" disabled={disabled}>
          Checkout
        </button>
      </Link>
    );
  }

  /**
   * @memberof Cart
   * @returns {JSX} Order Item
   */
  renderOrderSummary = () => {
    const {
      order, handleQuantityInputChange, removeOrderItem
    } = this.props;

    return (
      <div className="order-summary">
        {order.map(item => (
          <div key={item.id}>
            <p>
              <input type="number" className="order-input" value={item.quantity} min="1" onChange={e => handleQuantityInputChange(e, item.id)} />
            </p>
            <p data-tip={item.title} data-for="cart-tooltip">{item.title}</p>
            <p>{item.price}</p>
            <LinkBtn className="remove-order" dataTip="Remove Item?" data-for="cart-tooltip" clickHandler={() => removeOrderItem(item.id)}>&times;</LinkBtn>
            <ReactTooltip place="top" type="dark" effect="solid" className="tooltip" id="cart-tooltip" />
          </div>
        ))}
        <div className="order-amount">
          <div>
            <p>Total:</p>
            <h2>
&#8358;
              {calculateCashEarnedFromOrder(order)}
            </h2>
          </div>
        </div>
      </div>
    );
  }

  /**
   * @memberof Cart
   * @returns {JSX} Cart Toggle
   */
  renderCartToggle = () => {
    const { order } = this.props;
    const isShopOpen = checkShopOpen();

    return (
      <div className="side-cart-toggle d-none-xl">
        <span aria-hidden="true" id="cart-toggler" onClick={this.toggleCart}>
          <div className="text-center">
            <img src="/img/cart.png" alt="cart" />
            {' '}
            &nbsp;
            <sup>{order.length}</sup>
          </div>
          <p className="text-center">
            Total: &#8358;
            {calculateCashEarnedFromOrder(order)}
          </p>
        </span>
        {isShopOpen
          && (
          <span className="cart-btn">
            {this.renderCheckoutBtn()}
          </span>
          )}
      </div>
    );
  }

  /**
   * @memberof Cart
   * @returns {JSX} Cart Component
   */
  render() {
    const { order } = this.props;
    const isShopOpen = checkShopOpen();

    return (
      <Fragment>
        {this.renderCartToggle()}
        <div className={`side-cart ${this.state.show ? 'show' : null}`} id="side-cart">
          <div className="cart">
            <div className="cart-header">
              <h2>Your Basket</h2>
              <CloseIcon divClass="d-none-xl" btnID="side-cart-close-icon" clickHandler={this.toggleCart} />
            </div>
            <hr />
            {order.length === 0
              && (
              <div className="empty-cart">
                <p>Your Basket is Empty</p>
              </div>
              )}
            {order.length !== 0 && this.renderOrderSummary()}
            <br />
            {isShopOpen && this.renderCheckoutBtn()}
          </div>
        </div>
      </Fragment>
    );
  }
}

export default Cart;
