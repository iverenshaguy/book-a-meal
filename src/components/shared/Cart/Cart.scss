.side-cart h2 {
  text-align: center;
}

.side-cart .order-summary,
.side-cart .order-amount:not(.admin-order-total),
.side-cart .customer-details,
.side-cart .order-misc {
  border-bottom: 1px solid #ddd;
}

.side-cart .order-summary>div {
  display: flex;
}

.side-cart .order-summary>div {
  justify-content: center;
  align-items: center;
}

.side-cart .order-summary>div p:nth-child(1),
.side-cart .order-summary>div p:nth-child(3),
.side-cart .order-summary>div a,
.side-cart .order-summary>div .link-btn {
  flex-grow: 1;
}

.side-cart .order-summary>div p:nth-child(2) {
  font-size: 0.8em;

  @media (min-width: 1200px) {
    width: 50%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
}

.side-cart .order-summary input[type=number] {
  @media (max-width: 1199.98px) {
    border: 1px solid #333;
  }
}

.side-cart .order-summary>div a {
  padding-left: 5px;
}

.side-cart .order-summary>div select {
  flex-grow: 2;
}

.side-cart .order-summary>div p:nth-child(3),
.side-cart .order-summary>div a,
.side-cart .order-summary>div .link-btn {
  text-align: right;
}

.side-cart .order-amount>div {
  display: flex;
  justify-content: space-between;
  width: 100%;
}

.side-cart .order-amount>div:last-child {
  display: flex;
  align-items: center;
}