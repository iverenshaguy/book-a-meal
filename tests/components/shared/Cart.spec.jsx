import React from 'react';
import moment from 'moment';
import Cart from '../../../src/components/shared/Cart';
import { order } from '../../setup/mockData';

const { now } = Date;
const currentDay = moment().format('YYYY-MM-DD');

describe('Cart', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => new Date(currentDay).getTime() + (60 * 60 * 13 * 1000));
  });

  afterAll(() => {
    Date.now = now;
  });

  it('should render the Cart component correctly', () => {
    const shallowWrapper = shallow(<Cart
      order={order}
      handleQuantityInputChange={jest.fn()}
      removeOrderItem={jest.fn()}
    />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should render a message when there is no order', () => {
    const shallowWrapper = shallow(<Cart
      order={[]}
      handleQuantityInputChange={jest.fn()}
      removeOrderItem={jest.fn()}
    />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
    expect(shallowWrapper.find('div.empty-cart>p').text()).toEqual('Your Basket is Empty');
  });

  it('should toggle the cart when the cart toggler is clicked', () => {
    const wrapper = mount(<Cart
      order={order}
      handleQuantityInputChange={jest.fn()}
      removeOrderItem={jest.fn()}
    />, rrcMock.get());

    wrapper.find('#cart-toggler').simulate('click');

    expect(wrapper.state().show).toBeTruthy();
  });

  it('should show the checkout button when the shop is open', () => {
    const wrapper = mount(<Cart
      order={order}
      handleQuantityInputChange={jest.fn()}
      removeOrderItem={jest.fn()}
    />, rrcMock.get());

    expect(wrapper.find('.checkout-btn').length).toEqual(2); // 2 buttons for differet screen sizes
  });

  it('should hide the checkout button whenthe shop is closed', () => {
    const dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => new Date(currentDay).getTime() + (60 * 60 * 18 * 1000));
    const wrapper = mount(<Cart
      order={order}
      handleQuantityInputChange={jest.fn()}
      removeOrderItem={jest.fn()}
    />, rrcMock.get());

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.checkout-btn').length).toEqual(0);

    dateNowSpy.mockReset();
    dateNowSpy.mockRestore();
  });
});
