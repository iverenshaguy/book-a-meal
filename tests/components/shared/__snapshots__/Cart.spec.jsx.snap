// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should hide the checkout button whenthe shop is closed 1`] = `
<Cart
  handleQuantityInputChange={[MockFunction]}
  order={
    Array [
      Object {
        "id": "81211c24-51c0-46ec-b1e0-18db55880958",
        "price": "1500.00",
        "quantity": 1,
        "title": "Jollof Rice, Beef and Plantain",
      },
    ]
  }
  removeOrderItem={[MockFunction]}
>
  <div
    className="side-cart-toggle d-none-xl"
  >
    <span
      aria-hidden="true"
      id="cart-toggler"
      onClick={[Function]}
    >
      <div
        className="text-center"
      >
        <img
          alt="cart"
          src="/img/cart.png"
        />
         
         
        <sup>
          1
        </sup>
      </div>
      <p
        className="text-center"
      >
        Total: ₦
        1500
      </p>
    </span>
  </div>
  <div
    className="side-cart null"
    id="side-cart"
  >
    <div
      className="cart"
    >
      <div
        className="cart-header"
      >
        <h2>
          Your Basket
        </h2>
        <CloseIcon
          btnID="side-cart-close-icon"
          clickHandler={[Function]}
          divClass="d-none-xl"
        >
          <div
            className="close d-none-xl"
          >
            <Link
              aria-hidden="true"
              className={null}
              clickHandler={[Function]}
              dataTip={null}
              id="side-cart-close-icon"
            >
              <button
                className="link-btn null"
                data-tip={null}
                id="side-cart-close-icon"
                onClick={[Function]}
                type="button"
              >
                ×
              </button>
            </Link>
          </div>
        </CloseIcon>
      </div>
      <hr />
      <div
        className="order-summary"
      >
        <div
          key="81211c24-51c0-46ec-b1e0-18db55880958"
        >
          <p>
            <input
              className="order-input"
              min="1"
              onChange={[Function]}
              type="number"
              value={1}
            />
          </p>
          <p
            data-for="cart-tooltip"
            data-tip="Jollof Rice, Beef and Plantain"
          >
            Jollof Rice, Beef and Plantain
          </p>
          <p>
            1500.00
          </p>
          <Link
            className="remove-order"
            clickHandler={[Function]}
            data-for="cart-tooltip"
            dataTip="Remove Item?"
            id={null}
          >
            <button
              className="link-btn remove-order"
              data-tip="Remove Item?"
              id={null}
              onClick={[Function]}
              type="button"
            >
              ×
            </button>
          </Link>
          <ReactTooltip
            className="tooltip"
            clickable={false}
            effect="solid"
            id="cart-tooltip"
            insecure={true}
            place="top"
            resizeHide={true}
            type="dark"
            wrapper="div"
          >
            <div
              className="__react_component_tooltip place-top type-dark"
              data-id="tooltip"
              id="cart-tooltip"
            />
          </ReactTooltip>
        </div>
        <div
          className="order-amount"
        >
          <div>
            <p>
              Total:
            </p>
            <h2>
              ₦
              1500
            </h2>
          </div>
        </div>
      </div>
      <br />
    </div>
  </div>
</Cart>
`;

exports[`Cart should render a message when there is no order 1`] = `
<Fragment>
  <div
    className="side-cart-toggle d-none-xl"
  >
    <span
      aria-hidden="true"
      id="cart-toggler"
      onClick={[Function]}
    >
      <div
        className="text-center"
      >
        <img
          alt="cart"
          src="/img/cart.png"
        />
         
         
        <sup>
          0
        </sup>
      </div>
      <p
        className="text-center"
      >
        Total: ₦
        0
      </p>
    </span>
    <span
      className="cart-btn"
    >
      <Link
        replace={false}
        to="/order-review"
      >
        <button
          className="btn btn-pri btn-block checkout-btn"
          disabled="disabled"
          type="button"
        >
          Checkout
        </button>
      </Link>
    </span>
  </div>
  <div
    className="side-cart null"
    id="side-cart"
  >
    <div
      className="cart"
    >
      <div
        className="cart-header"
      >
        <h2>
          Your Basket
        </h2>
        <CloseIcon
          btnID="side-cart-close-icon"
          clickHandler={[Function]}
          divClass="d-none-xl"
        />
      </div>
      <hr />
      <div
        className="empty-cart"
      >
        <p>
          Your Basket is Empty
        </p>
      </div>
      <br />
      <Link
        replace={false}
        to="/order-review"
      >
        <button
          className="btn btn-pri btn-block checkout-btn"
          disabled="disabled"
          type="button"
        >
          Checkout
        </button>
      </Link>
    </div>
  </div>
</Fragment>
`;

exports[`Cart should render the Cart component correctly 1`] = `
<Fragment>
  <div
    className="side-cart-toggle d-none-xl"
  >
    <span
      aria-hidden="true"
      id="cart-toggler"
      onClick={[Function]}
    >
      <div
        className="text-center"
      >
        <img
          alt="cart"
          src="/img/cart.png"
        />
         
         
        <sup>
          1
        </sup>
      </div>
      <p
        className="text-center"
      >
        Total: ₦
        1500
      </p>
    </span>
    <span
      className="cart-btn"
    >
      <Link
        replace={false}
        to="/order-review"
      >
        <button
          className="btn btn-pri btn-block checkout-btn"
          disabled={null}
          type="button"
        >
          Checkout
        </button>
      </Link>
    </span>
  </div>
  <div
    className="side-cart null"
    id="side-cart"
  >
    <div
      className="cart"
    >
      <div
        className="cart-header"
      >
        <h2>
          Your Basket
        </h2>
        <CloseIcon
          btnID="side-cart-close-icon"
          clickHandler={[Function]}
          divClass="d-none-xl"
        />
      </div>
      <hr />
      <div
        className="order-summary"
      >
        <div
          key="81211c24-51c0-46ec-b1e0-18db55880958"
        >
          <p>
            <input
              className="order-input"
              min="1"
              onChange={[Function]}
              type="number"
              value={1}
            />
          </p>
          <p
            data-for="cart-tooltip"
            data-tip="Jollof Rice, Beef and Plantain"
          >
            Jollof Rice, Beef and Plantain
          </p>
          <p>
            1500.00
          </p>
          <Link
            className="remove-order"
            clickHandler={[Function]}
            data-for="cart-tooltip"
            dataTip="Remove Item?"
            id={null}
          >
            ×
          </Link>
          <ReactTooltip
            className="tooltip"
            clickable={false}
            effect="solid"
            id="cart-tooltip"
            insecure={true}
            place="top"
            resizeHide={true}
            type="dark"
            wrapper="div"
          />
        </div>
        <div
          className="order-amount"
        >
          <div>
            <p>
              Total:
            </p>
            <h2>
              ₦
              1500
            </h2>
          </div>
        </div>
      </div>
      <br />
      <Link
        replace={false}
        to="/order-review"
      >
        <button
          className="btn btn-pri btn-block checkout-btn"
          disabled={null}
          type="button"
        >
          Checkout
        </button>
      </Link>
    </div>
  </div>
</Fragment>
`;
